File: typescript/packages/core/src/core/gmail/readdir.ts
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

import type { GmailAccessor } from '../../accessor/gmail.ts'
import { IndexEntry } from '../../cache/index/config.ts'
import type { IndexCacheStore } from '../../cache/index/store.ts'
import { PathSpec } from '../../types.ts'
import { listLabels } from './labels.ts'
import type { GmailMessageRaw } from './messages.ts'
import { extractAttachments, extractHeader, getMessageRaw, listMessages } from './messages.ts'
import { stripSlash } from '../../util/slash.ts'

const TITLE_MAX = 80
const UNSAFE = /[^\w\s\-.]/g
const MULTI_UNDERSCORE = /_+/g

export function sanitize(text: string): string {
  if (text.trim() === '') return 'No_Subject'
  let cleaned = text.replace(UNSAFE, '_').replace(/ /g, '_')
  cleaned = cleaned.replace(MULTI_UNDERSCORE, '_').replace(/^_+|_+$/g, '')
  if (cleaned.length > TITLE_MAX) cleaned = `${cleaned.slice(0, TITLE_MAX - 3)}...`
  return cleaned
}

function msgFilename(subject: string, msgId: string): string {
  return `${sanitize(subject)}__${msgId}.gmail.json`
}

function dateFromInternal(internalDate: string | undefined): string {
  if (internalDate === undefined) return '1970-01-01'
  const ts = Number.parseInt(internalDate, 10)
  if (!Number.isFinite(ts)) return '1970-01-01'
  const d = new Date(ts)
  const yyyy = d.getUTCFullYear().toString().padStart(4, '0')
  const mm = (d.getUTCMonth() + 1).toString().padStart(2, '0')
  const dd = d.getUTCDate().toString().padStart(2, '0')
  return `${yyyy}-${mm}-${dd}`
}

function enoent(p: string): Error {
  const e = new Error(`ENOENT: ${p}`) as Error & { code: string }
  e.code = 'ENOENT'
  return e
}

export async function readdir(
  accessor: GmailAccessor,
  path: PathSpec,
  index?: IndexCacheStore,
): Promise<string[]> {
  const prefix = path.prefix
  const raw = path.pattern !== null ? path.directory : path.original
  let p = raw
  if (prefix !== '' && p.startsWith(prefix)) p = p.slice(prefix.length) || '/'
  const key = stripSlash(p)
  const virtualKey = key !== '' ? `${prefix}/${key}` : prefix !== '' ? prefix : '/'
  const parts = key === '' ? [] : key.split('/')
  const depth = parts.length

  if (depth === 0) {
    if (index !== undefined) {
      const cached = await index.listDir(virtualKey)
      if (cached.entries !== undefined && cached.entries !== null) return cached.entries
    }
    const labels = await listLabels(accessor.tokenManager)
    const entries: [string, IndexEntry][] = []
    for (const lb of labels) {
      const name = lb.type === 'system' ? lb.id : (lb.name ?? lb.id)
      const entry = new IndexEntry({
        id: lb.id,
        name,
        resourceType: 'gmail/label',
        vfsName: name,
      })
      entries.push([name, entry])
    }
    if (index !== undefined) await index.setDir(virtualKey, entries)
    return entries.map(([name]) => `${prefix}/${name}`)
  }

  if (depth === 1) {
    const labelName = parts[0] ?? ''
    if (index !== undefined) {
      const cached = await index.listDir(virtualKey)
      if (cached.entries !== undefined && cached.entries !== null) return cached.entries
    }
    if (index === undefined) throw enoent(path.original)
    const labelKey = prefix !== '' ? `${prefix}/${labelName}` : `/${labelName}`
    let result = await index.get(labelKey)
    if (result.entry === undefined || result.entry === null) {
      try {
        const root = new PathSpec({
          original: prefix !== '' ? prefix : '/',
          directory: prefix !== '' ? prefix : '/',
          prefix,
        })
        await readdir(accessor, root, index)
        result = await index.get(labelKey)
      } catch {
        // ignore — falls through to ENOENT below
      }
    }
    if (result.entry === undefined || result.entry === null) throw enoent(path.original)
    const labelId = result.entry.id
    const msgIds = await listMessages(accessor.tokenManager, { labelId, maxResults: 50 })
    const dateGroups = new Map<string, GmailMessageRaw[]>()
    for (const m of msgIds) {
      const mid = m.id
      const rawMsg = await getMessageRaw(accessor.tokenManager, mid)
      const dateStr = dateFromInternal(rawMsg.internalDate)
      let bucket = dateGroups.get(dateStr)
      if (bucket === undefined) {
        bucket = []
        dateGroups.set(dateStr, bucket)
      }
      bucket.push(rawMsg)
    }
    const sortedDates = [...dateGroups.keys()].sort().reverse()
    const dateEntries: [string, IndexEntry][] = []
    for (const dateStr of sortedDates) {
      const dateEntry = new IndexEntry({
        id: dateStr,
        name: dateStr,
        resourceType: 'gmail/date',
        vfsName: dateStr,
      })
      dateEntries.push([dateStr, dateEntry])
      const msgEntries: [string, IndexEntry][] = []
      for (const rawMsg of dateGroups.get(dateStr) ?? []) {
        const mid = rawMsg.id ?? ''
        const headers = rawMsg.payload?.headers ?? []
        const subject = extractHeader(headers, 'Subject') || 'No Subject'
        const filename = msgFilename(subject, mid)
        const msgEntry = new IndexEntry({
          id: mid,
          name: subject,
          resourceType: 'gmail/message',
          vfsName: filename,
          size: rawMsg.sizeEstimate ?? null,
        })
        msgEntries.push([filename, msgEntry])
        const attachments = extractAttachments(rawMsg.payload)
        if (attachments.length > 0) {
          const attDirName = filename.replace('.gmail.json', '')
          const attDirEntry = new IndexEntry({
            id: mid,
            name: attDirName,
            resourceType: 'gmail/attachment_dir',
            vfsName: attDirName,
          })
          msgEntries.push([attDirName, attDirEntry])
          const attEntries: [string, IndexEntry][] = []
          for (const att of attachments) {
            const attEntry = new IndexEntry({
              id: att.attachmentId,
              name: att.filename,
              resourceType: 'gmail/attachment',
              vfsName: att.filename,
              size: att.size,
            })
            attEntries.push([att.filename, attEntry])
          }
          const attDirVKey = `${virtualKey}/${dateStr}/${attDirName}`
          await index.setDir(attDirVKey, attEntries)
        }
      }
      const dateVKey = `${virtualKey}/${dateStr}`
      await index.setDir(dateVKey, msgEntries)
    }
    await index.setDir(virtualKey, dateEntries)
    return dateEntries.map(([name]) => `${prefix}/${key}/${name}`)
  }

  if (depth === 2 || depth === 3) {
    if (index === undefined) throw enoent(path.original)
    let cached = await index.listDir(virtualKey)
    if (cached.entries !== undefined && cached.entries !== null) return cached.entries
    const labelPath = prefix !== '' ? `${prefix}/${parts[0] ?? ''}` : `/${parts[0] ?? ''}`
    const labelSpec = new PathSpec({ original: labelPath, directory: labelPath, prefix })
    await readdir(accessor, labelSpec, index)
    cached = await index.listDir(virtualKey)
    if (cached.entries !== undefined && cached.entries !== null) return cached.entries
    throw enoent(path.original)
  }

  throw enoent(path.original)
}

File: python/mirage/core/gmail/readdir.py
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

import logging
import re
from datetime import datetime, timezone

from mirage.accessor.gmail import GmailAccessor
from mirage.cache.index import IndexCacheStore, IndexEntry
from mirage.core.gmail.date_query import date_dir_to_gmail_query
from mirage.core.gmail.labels import list_labels
from mirage.core.gmail.messages import (_extract_attachments, _extract_header,
                                        get_message_raw, list_messages)
from mirage.core.google.drive import GoogleFileSuffix
from mirage.types import PathSpec
from mirage.utils.errors import enoent

logger = logging.getLogger(__name__)


def is_dir_name(child: str) -> bool:
    # readdir emits only label/date dirs and rendered *.gmail.json files.
    return not child.endswith(GoogleFileSuffix.GMAIL.value)


TITLE_MAX = 80
UNSAFE = re.compile(r"[^\w\s\-.]")
MULTI_UNDERSCORE = re.compile(r"_+")


def _sanitize(text: str) -> str:
    if not text.strip():
        return "No_Subject"
    cleaned = UNSAFE.sub("_", text).replace(" ", "_")
    cleaned = MULTI_UNDERSCORE.sub("_", cleaned).strip("_")
    if len(cleaned) > TITLE_MAX:
        cleaned = cleaned[:TITLE_MAX - 3] + "..."
    return cleaned


def _msg_filename(subject: str, msg_id: str) -> str:
    return f"{_sanitize(subject)}__{msg_id}.gmail.json"


def _attach_dir_name(subject: str, msg_id: str) -> str:
    return f"{_sanitize(subject)}__{msg_id}"


def _attachment_filename(_attachment_id: str, filename: str) -> str:
    return filename or "file"


def _date_from_internal(internal_date: str) -> str:
    ts = int(internal_date) / 1000
    return datetime.fromtimestamp(ts, tz=timezone.utc).strftime("%Y-%m-%d")


async def _build_date_groups(
    accessor: GmailAccessor,
    msg_ids: list[dict],
    index: IndexCacheStore | None,
    virtual_key: str,
    write_dates: bool,
) -> list[tuple[str, IndexEntry]]:
    date_groups: dict[str, list[dict]] = {}
    for m in msg_ids:
        mid = m["id"]
        raw = await get_message_raw(accessor.token_manager, mid)
        internal_date = raw.get("internalDate", "0")
        date_str = _date_from_internal(internal_date)
        date_groups.setdefault(date_str, []).append(raw)
    date_entries: list[tuple[str, IndexEntry]] = []
    for date_str in sorted(date_groups.keys(), reverse=True):
        date_entry = IndexEntry(
            id=date_str,
            name=date_str,
            resource_type="gmail/date",
            vfs_name=date_str,
        )
        date_entries.append((date_str, date_entry))
        date_children: list[tuple[str, IndexEntry]] = []
        for raw in date_groups[date_str]:
            mid = raw["id"]
            headers = raw.get("payload", {}).get("headers", [])
            subject = _extract_header(headers, "Subject") or "No Subject"
            filename = _msg_filename(subject, mid)
            msg_entry = IndexEntry(
                id=mid,
                name=subject,
                resource_type="gmail/message",
                vfs_name=filename,
                size=raw.get("sizeEstimate"),
            )
            date_children.append((filename, msg_entry))
            attachments = _extract_attachments(raw.get("payload", {}))
            if attachments:
                att_dir = _attach_dir_name(subject, mid)
                att_dir_entry = IndexEntry(
                    id=mid,
                    name=att_dir,
                    resource_type="gmail/attachment_dir",
                    vfs_name=att_dir,
                )
                date_children.append((att_dir, att_dir_entry))
                att_entries: list[tuple[str, IndexEntry]] = []
                for att in attachments:
                    att_name = _attachment_filename(att["attachment_id"],
                                                    att["filename"])
                    att_entry = IndexEntry(
                        id=att["attachment_id"],
                        name=att["filename"],
                        resource_type="gmail/attachment",
                        vfs_name=att_name,
                        size=att["size"],
                    )
                    att_entries.append((att_name, att_entry))
                att_vkey = virtual_key + "/" + date_str + "/" + att_dir
                if index is not None and write_dates:
                    await index.set_dir(att_vkey, att_entries)
        if index is not None and write_dates:
            await index.set_dir(virtual_key + "/" + date_str, date_children)
    return date_entries


async def readdir(
    accessor: GmailAccessor,
    path: PathSpec,
    index: IndexCacheStore = None,
) -> list[str]:
    if isinstance(path, str):
        path = PathSpec(original=path, directory=path)
    virtual = path.original
    if isinstance(path, PathSpec):
        prefix = path.prefix
        path = path.directory if path.pattern else path.original
    if prefix and path.startswith(prefix):
        rest = path[len(prefix):]
        if prefix.endswith("/") or rest == "" or rest.startswith("/"):
            path = rest or "/"
    key = path.strip("/")
    virtual_key = prefix + "/" + key if key else prefix or "/"
    parts = key.split("/") if key else []
    depth = len(parts)

    if depth == 0:
        if index is not None:
            cached = await index.list_dir(virtual_key)
            if cached.entries is not None:
                return cached.entries
        labels = await list_labels(accessor.token_manager)
        entries = []
        for lb in labels:
            if lb.get("type") == "system":
                name = lb["id"]
            else:
                name = lb.get("name", lb["id"])
            entry = IndexEntry(
                id=lb["id"],
                name=name,
                resource_type="gmail/label",
                vfs_name=name,
            )
            entries.append((name, entry))
        if index is not None:
            await index.set_dir(virtual_key, entries)
        return [f"{prefix}/{name}" for name, _ in entries]

    if depth == 1:
        label_name = parts[0]
        if index is not None:
            cached = await index.list_dir(virtual_key)
            if cached.entries is not None:
                return cached.entries
        if index is None:
            raise enoent(virtual)
        label_key = prefix + "/" + label_name if prefix else "/" + label_name
        result = await index.get(label_key)
        if result.entry is None:
            # Auto-bootstrap: populate label index.
            try:
                root = PathSpec(
                    original=prefix or "/",
                    directory=prefix or "/",
                    prefix=prefix,
                )
                await readdir(accessor, root, index)
                result = await index.get(label_key)
            except Exception as e:
                logger.debug(
                    "gmail readdir: bootstrap failed for %s: %s",
                    label_key,
                    e,
                )
        if result.entry is None:
            raise enoent(virtual)
        label_id = result.entry.id
        msg_ids = await list_messages(
            accessor.token_manager,
            label_id=label_id,
            max_results=50,
        )
        date_entries = await _build_date_groups(
            accessor,
            msg_ids,
            index,
            virtual_key,
            write_dates=True,
        )
        if index is not None:
            await index.set_dir(virtual_key, date_entries)
        return [f"{prefix}/{key}/{name}" for name, _ in date_entries]

    if depth == 2:
        if index is None:
            raise enoent(virtual)
        cached = await index.list_dir(virtual_key)
        if cached.entries is not None:
            return cached.entries
        label_name = parts[0]
        date_str = parts[1]
        date_query = date_dir_to_gmail_query(date_str)
        if date_query is not None:
            label_key = (prefix + "/" + label_name if prefix else "/" +
                         label_name)
            label_result = await index.get(label_key)
            if label_result.entry is None:
                try:
                    root = PathSpec(
                        original=prefix or "/",
                        directory=prefix or "/",
                        prefix=prefix,
                    )
                    await readdir(accessor, root, index)
                    label_result = await index.get(label_key)
                except Exception as e:
                    logger.debug(
                        "gmail readdir: date-dir bootstrap failed for %s: %s",
                        label_key,
                        e,
                    )
            if label_result.entry is not None:
                label_id = label_result.entry.id
                label_vkey = label_key
                msg_ids = await list_messages(
                    accessor.token_manager,
                    label_id=label_id,
                    query=date_query,
                    max_results=50,
                )
                await _build_date_groups(
                    accessor,
                    msg_ids,
                    index,
                    label_vkey,
                    write_dates=True,
                )
                if not msg_ids:
                    await index.set_dir(virtual_key, [])
                cached = await index.list_dir(virtual_key)
                if cached.entries is not None:
                    return cached.entries
                raise enoent(virtual)
        label_path = prefix + "/" + parts[0] if prefix else "/" + parts[0]
        await readdir(accessor,
                      PathSpec.from_str_path(label_path, prefix=prefix), index)
        cached = await index.list_dir(virtual_key)
        if cached.entries is not None:
            return cached.entries
        raise enoent(virtual)

    if depth == 3:
        if index is None:
            raise enoent(virtual)
        cached = await index.list_dir(virtual_key)
        if cached.entries is not None:
            return cached.entries
        date_path = (prefix + "/" + parts[0] + "/" +
                     parts[1] if prefix else "/" + parts[0] + "/" + parts[1])
        await readdir(accessor, PathSpec.from_str_path(date_path,
                                                       prefix=prefix), index)
        cached = await index.list_dir(virtual_key)
        if cached.entries is not None:
            return cached.entries
        raise enoent(virtual)

    raise enoent(virtual)
